### ISSUE TITLE
Is This Preprocessing Pipeline Suitable for H&E-Stained RGB Input?

### ISSUE THREAD

**Issue by vadori** [vanvalenlab/cellSAM#74]
Hi,

I’m working with RGB histology images, specifically H&E-stained images with size 256x256, and would like to verify if the preprocessing function below is appropriate for using the model with the function segment_cellular_image:

```
def process_image_array(image_array):
    # Ensure input is float32 for calculations
    img = image_array.astype(np.float32)

    # Get original size
    h, w, _ = img.shape
    desired_size = 512

    # Compute average across channels before normalization
    avg = image_array.mean(axis=2)

    # Normalize average to [0, 1] using min-max scaling
    avg_normalized = (avg - avg.min()) / (avg.max() - avg.min() + 1e-8)

    # Create output RGB image with only blue channel set
    output = np.zeros((*avg_normalized.shape, 3), dtype=np.float32)
    output[..., 2] = avg_normalized

    # Pad to 512x512 (centered padding)
    h_new, w_new = output.shape[:2]
    pad_h = max(0, desired_size - h_new)
    pad_w = max(0, desired_size - w_new)
    pad_top = pad_h // 2
    pad_bottom = pad_h - pad_top
    pad_left = pad_w // 2
    pad_right = pad_w - pad_left

    output_padded = np.pad(
        output,
        ((pad_top, pad_bottom), (pad_left, pad_right), (0, 0)),
        mode='constant',
        constant_values=0
    )

    # Convert to (3, 512, 512)
    output_ch_first = np.transpose(output_padded, (2, 0, 1))

    return output_ch_first
```

In summary, I am:

- Converting the input RGB image to float32.
- Computing the average intensity across RGB channels.
- Min-max normalizing the average to the [0, 1] range.
- Creating an output RGB image where only the blue channel is set to the normalized average.
- Padding the image (centered) to 512×512 if necessary.
- Returning the final image in (3, 512, 512) channel-first format for model input.

I’d appreciate any feedback on whether this pipeline is suitable for RGB, particularly H&E-stained, histology images.

**Comment by rossbar** [comment]
I'd recommend to try `cellsam_pipeline` first with the raw image without any custom preprocessing. Any necessary image processing (which essentially amounts to casting to float and normalizing to the range [0, 1]) already happens internally within `cellsam_pipeline`, so there's no need to do any additional preprocessing beforehand.

As to whether this posted preprocessing function will work... I expect so, but perhaps not as well as the raw image as some of the operations (e.g. the averaging) likely move the input data further away from the training corpus. No real idea though, just my opinion!

**Comment by vadori** [comment]
Hi @rossbar,

Thank you for your response. I’ll go ahead and test the `cellsam_pipeline`. I based the steps above on a passage from the Supplementary Materials of the paper, aiming to minimize the discrepancy between the RGB images used during training and those I’m using for inference. As for the averaging step I should probably remove it. It was the CellPose-SAM paper that led to some confusion in that regard.

From https://arxiv.org/pdf/2311.11004:
<img width="735" height="900" alt="Image" src="https://github.com/user-attachments/assets/f59012d5-ee61-49fe-938a-de1c27aa9c67" />

From https://www.biorxiv.org/content/10.1101/2025.04.28.651001v1.full.pdf:
<img width="332" height="313" alt="Image" src="https://github.com/user-attachments/assets/323d04de-5d5a-46ff-95c1-baadf7ec0c97" />

**Comment by vadori** [comment]
Another attempt, everything the same except for the padding (removed). I'd say the last option is the best.

<img width="1548" height="472" alt="Image" src="https://github.com/user-attachments/assets/022a9c30-1ebc-4904-8ad7-e574e437bf31" />

<img width="1548" height="472" alt="Image" src="https://github.com/user-attachments/assets/9c97062c-00cc-431e-bdc4-80ab4414b0a7" />

<img width="1548" height="472" alt="Image" src="https://github.com/user-attachments/assets/bbb2a174-5ec3-4ce9-89f2-8d20cff07a8b" />

<img width="1548" height="472" alt="Image" src="https://github.com/user-attachments/assets/a8b683b7-6114-448f-9833-2ccc52968ba4" />

<img width="1548" height="472" alt="Image" src="https://github.com/user-attachments/assets/ef5e36fc-d765-4134-bb3b-5a06f2e476d1" />



```
import numpy as np
import cv2
import matplotlib.pyplot as plt
from cellSAM import segment_cellular_image, cellsam_pipeline

def center_crop_mask(mask, crop_size=256):
    h, w = mask.shape
    top = (h - crop_size) // 2
    left = (w - crop_size) // 2
    return mask[top:top + crop_size, left:left + crop_size]

def process_image_array(image_array, bluechannel=False):
    # Ensure input is float32 for calculations
    img = image_array.astype(np.float32)

    # Get original size
    h, w, _ = img.shape
    desired_size = 512

    if not bluechannel:
        output = (img - img.min(axis=(0, 1))) / (img.max(axis=(0, 1)) - img.min(axis=(0, 1)) + 1e-8)
    
    else:

        # Compute average across channels before normalization
        avg = img.mean(axis=2)

        # Normalize average to [0, 1] using min-max scaling
        avg_normalized = (avg - avg.min()) / (avg.max() - avg.min() + 1e-8)  # small epsilon to avoid divide by zero

        # Create output RGB image with only blue channel set
        output = np.zeros((*avg_normalized.shape, 3), dtype=np.float32)
        output[..., 2] = avg_normalized

    # # Pad to 512x512 (centered padding)
    # h_new, w_new = output.shape[:2]
    # pad_h = max(0, desired_size - h_new)
    # pad_w = max(0, desired_size - w_new)
    # pad_top = pad_h // 2
    # pad_bottom = pad_h - pad_top
    # pad_left = pad_w // 2
    # pad_right = pad_w - pad_left

    # output_padded = np.pad(
    #     output,
    #     ((pad_top, pad_bottom), (pad_left, pad_right), (0, 0)),
    #     mode='constant',
    #     constant_values=0
    # )

    output_padded = output

    # # Convert to (3, 512, 512)
    # output_ch_first = np.transpose(output_padded, (2, 0, 1))

    # return output_ch_first
    
    return output_padded

def upsample_rgb_image(image):
    return cv2.resize(image, (512, 512), interpolation=cv2.INTER_LINEAR)

def downsample_label_mask(mask, target_size):
    return cv2.resize(mask, target_size, interpolation=cv2.INTER_NEAREST)

bgr_paths = ['D:/NucleiSegmentation/Projects/cytoark/datasets/conic/20x/256x256/image/crag_60-0021.png',
'D:/NucleiSegmentation/Projects/cytoark/datasets/conic/20x/256x256/image/crag_9-0032.png',
'D:/NucleiSegmentation/Projects/cytoark/datasets/pannuke/40x/256x256/image/pannuke_bladder_fold3_1583.png',
'D:/NucleiSegmentation/Projects/cytoark/datasets/pannuke/40x/256x256/image/pannuke_colon_fold1_2619.png',
'D:/NucleiSegmentation/Projects/cytoark/datasets/pannuke/40x/256x256/image/pannuke_colon_fold1_2609.png'
]

for i in range(len(bgr_paths)):
    bgr_image = cv2.imread(bgr_paths[i])
    rgb_image = cv2.cvtColor(bgr_image, cv2.COLOR_BGR2RGB)
    
    if i < 2:
        rgb_image = upsample_rgb_image(rgb_image)

    #segment
    mask1 = cellsam_pipeline(rgb_image)
    mask2 = segment_cellular_image(process_image_array(rgb_image), device='cuda')[0]
    mask3 = segment_cellular_image(process_image_array(rgb_image, bluechannel=True), device='cuda')[0]

    if i < 2:
        mask2 = downsample_label_mask(mask2, (256, 256))  # from (512, 512) to (256, 256)
        mask3 = downsample_label_mask(mask3, (256, 256))  # from (512, 512) to (256, 256)

    mask2 = center_crop_mask(mask2)
    mask3 = center_crop_mask(mask3)

    plt.figure(dpi=300)
    plt.subplot(1,4,1)
    plt.imshow(rgb_image)
    plt.axis('off')
    plt.title('original image', fontsize=5)
    plt.subplot(1,4,2)
    plt.imshow(mask1)
    plt.axis('off')
    plt.title('cellsam_pipeline output', fontsize=5)
    plt.subplot(1,4,3)
    plt.imshow(mask2)
    plt.axis('off')
    plt.title('segment_cellular_image *\n process_image_array output', fontsize=5)
    plt.subplot(1,4,4)
    plt.imshow(mask3)
    plt.axis('off')
    plt.title('segment_cellular_image *\n process_image_array output\n - only blue channel', fontsize=5)
```

**Comment by rossbar** [comment]
Thanks for sharing @vadori ! In my non-expert opinion, your experiment setup and tested parameters seem reasonable. I leave it to one of the experts whether there are any finer details that warrant comment - LUK @ulisrael @damaggu @rdilip @m274d .

The only other comment I have is related to the versioning of the pre-trained model weights --- the pre-trained model weights available to this point are from the pre-release on bioarxiv. There are now a set of "official" pre-trained model weights that are properly versioned and slated for release with the publication of the paper. These model weights are currently available, though not yet merged into the development branch of the software - see #88 for details.

**Comment by dsatyam09** [comment]
hey just to confirm is the input size which the cell sam model takes is of 512*512. bze I have an H&E stained image of (9510, 9993, 3) so I was a little confused if reducing the dimension so much would lead to a really poor segmentation or should I rather implement to tile larger images (e.g. 9510×9993) into blocks (512, 1024 or some power-of-two) and run inference patch-by-patch, then stitch results.    I would really appreciate your advice on this @rossbar.

**Comment by rossbar** [comment]
Thanks for the follow-up @vadori and the ping @dsatyam09 . There are two separate issues here:

The first is @vadori 's original question: what format does cellsam expect for H&E (or similar stained) images. The answer to this question is: gray scale - in other words, multi-channel stained images should be collapsed to grayscale for best results. There is a [gallery example](https://vanvalenlab.github.io/cellSAM/auto_examples/plot_HandE.html#sphx-glr-auto-examples-plot-hande-py) which demonstrates the expected format for histology images. Note that the documentation recommends otherwise, so there is a definite issue with the documentation that needs to be fixed!

The second question regards image scale - @dsatyam09 your intuition is correct, you should **not** be rescaling images to fit a 512x512 array, but rather use `use_wsi=True` to activate under-the-hood tiling for large images.

**Comment by DarioS** [comment]
> multi-channel stained images should be collapsed to grayscale for best results

Neither the journal article nor the software user guide specify how this was done for model training. There may be multiple marker choices for a cell compartment and knowing exactly which one training used would probably help segmentation.
